### PR TITLE
fix(op-node): Fix Channel Out MaxSize Underflow

### PR DESCRIPTION
**Pull request by refcell**
**Description**

Fixes CLI-3649

Context in: #5186. In short, the `maxSize` parameter of the channel out's `OutputFrame` function is unchecked so the `23` frame overhead size will underflow, resulting in a nearly max `uint64` frame size when a `maxSize` parameter of < 23 is provided.

### CODE REVIEW

File: op-node/rollup/derive/channel_out_test.go [hunk 1]
```diff
@@ -29,6 +29,26 @@ func TestChannelOutAddBlock(t *testing.T) {
...
 }
 
+// TestOutputFrameSmallMaxSize tests that calling [OutputFrame] with a small
+// max size that is below the fixed frame size overhead of 23, will return
+// an error.
+func TestOutputFrameSmallMaxSize(t *testing.T) {
+	cout, err := NewChannelOut()
+	require.NoError(t, err)
+
+	// Set the channel out frame details
```
Comment (sebastianst) [inline op-node/rollup/derive/channel_out_test.go:39]: Nit: test also just works without setting this. And 0 is the default value of int types.

File: op-node/rollup/derive/channel_out_test.go [hunk 2]
```diff
@@ -29,6 +29,26 @@ func TestChannelOutAddBlock(t *testing.T) {
...
+	cout.id = ChannelID{0x01}
+	cout.frame = 0
+
+	// Call OutputFrame with the range of small max size values that err
+	w := bytes.NewBuffer([]byte{})
```
Comment (sebastianst) [inline op-node/rollup/derive/channel_out_test.go:44]:
```suggestion
	var w bytes.Buffer
```
And use &w

**Review by sebastianst — CHANGES_REQUESTED** [review]
2 small nits, looks good otherwise ✨

File: op-node/rollup/derive/channel_out_test.go (hunk 1 above)
Comment (refcell) [reply op-node/rollup/derive/channel_out_test.go:39]: yessir, just boilerplate from when i set it up - removing ✅

File: op-node/rollup/derive/channel_out.go [hunk 3]
```diff
@@ -150,6 +151,12 @@ func (co *ChannelOut) OutputFrame(w *bytes.Buffer, maxSize uint64) (uint16, erro
 		FrameNumber: uint16(co.frame),
 	}
 
+	// Error if the `maxSize` is too small to fit the fixed frame
+	// overhead as specified below.
+	if maxSize < 23 {
```
Comment (trianglesphere) [inline op-node/rollup/derive/channel_out.go:156]: 23 should be a named constant (something like FrameV0OverHeadSize)

File: op-node/rollup/derive/channel_out.go [hunk 4]
```diff
@@ -150,6 +151,12 @@ func (co *ChannelOut) OutputFrame(w *bytes.Buffer, maxSize uint64) (uint16, erro
 		FrameNumber: uint16(co.frame),
 	}
 
+	// Error if the `maxSize` is too small to fit the fixed frame
```
Comment (trianglesphere) [inline op-node/rollup/derive/channel_out.go:154]: Better to indicate in the godoc comment that `maxSize must >= FrameOverHeadSize` than have this comment.

File: op-batcher/batcher/channel_manager_test.go [hunk 5]
```diff
@@ -199,29 +192,41 @@ func TestClearChannelManager(t *testing.T) {
 	require.Empty(t, m.confirmedTransactions)
 
 	// Add a block to the channel manager
-	a, _ := derivetest.RandomL2Block(rng, 4)
+	lBlock := types.NewBlock(&types.Header{
```
Comment (trianglesphere) [inline op-batcher/batcher/channel_manager_test.go:195]: why was this changed?

**Review by trianglesphere — CHANGES_REQUESTED** [review]
There's a lot of nice cleanup changes in here, but those are unrelated to primary concern of the PR which is fixing the MaxSizeUnderflow. Those changes should be split out to a new PR.

File: op-batcher/batcher/channel_manager_test.go (hunk 5 above)
Comment (refcell) [reply op-batcher/batcher/channel_manager_test.go:195]: Ah it broke because the randomness would allow the block to be too small

File: op-node/rollup/derive/channel_out.go (hunk 3 above)
Comment (refcell) [reply op-node/rollup/derive/channel_out.go:156]: Good catch, thanks!

File: op-node/rollup/derive/channel_out.go [hunk 6]
```diff
@@ -14,9 +14,15 @@ import (
 	"github.com/ethereum/go-ethereum/rlp"
 )
 
+var ErrMaxFrameSizeTooSmall = errors.New("maxSize is too small to fit the fixed frame overhead")
 var ErrNotDepositTx = errors.New("first transaction in block is not a deposit tx")
 var ErrTooManyRLPBytes = errors.New("batch would cause RLP bytes to go over limit")
 
+// FrameV0OverHeadSize is the absolute minimum size of a frame.
+// This is the fixed overhead frame size,
+// calculated as follows: 16 + 2 + 4 + 1 = 23 bytes.
```
Comment (tynes) [inline op-node/rollup/derive/channel_out.go:23]: Unclear where those numbers come from but the math checks out

Comment (sebastianst) [reply op-node/rollup/derive/channel_out.go:23]: comes from the [Frame Format specs](https://github.com/ethereum-optimism/optimism/blob/develop/specs/derivation.md#frame-format)

Comment (sebastianst) [reply op-node/rollup/derive/channel_out.go:23]: maybe link in this comment.

File: op-batcher/batcher/channel_builder_test.go [hunk 7]
```diff
@@ -400,35 +399,26 @@ func TestOutputFrames(t *testing.T) {
...
-	// There should be one frame in the channel builder now
-	require.Equal(t, 1, cb.NumFrames())
+	// We should be able to output the frames
+	require.NoError(t, cb.OutputFrames())
 
-	// There should no longer be any ready bytes
-	readyBytes = cb.co.ReadyBytes()
-	require.Equal(t, 0, readyBytes)
+	// There should be many frames in the channel builder now
+	require.Greater(t, cb.NumFrames(), 1)
```
Comment (sebastianst) [inline op-batcher/batcher/channel_builder_test.go:421]: You could add to that test to actually get a frame and assert that its size it equal to `MaxFrameSize`.

File: op-batcher/batcher/channel_manager_test.go [hunk 8]
```diff
@@ -222,8 +213,7 @@ func TestClearChannelManager(t *testing.T) {
 		Number:     big.NewInt(1),
 		ParentHash: a.Hash(),
 	}, nil, nil, nil, nil)
-	err = m.AddL2Block(b)
-	require.NoError(t, err)
+	require.NoError(t, m.AddL2Block(b))
 	require.Equal(t, 1, len(m.blocks))
```
Comment (sebastianst) [inline op-batcher/batcher/channel_manager_test.go:217]: there's `require.Len`

**Review by sebastianst — CHANGES_REQUESTED** [review]
Looking good now! Would be nice to still see the remaining tests that got refactored also get a proper test name.

**Review by sebastianst — APPROVED** [review]

**Review by protolambda — APPROVED** [review]
